### PR TITLE
Resolve permissions errors

### LINKED ISSUE #30: Handle pushes to main branch
Someday I'll iron out all the bugs...

I tested #29 with pull requests but I never _merged_ them. If I had, I would have seen failures like https://github.com/uclahs-cds/pipeline-call-sSNV/actions/runs/8530035065/job/23367065681.

The workflow should gracefully handle checks that do not have an associated pull request.

### LINKED ISSUE #31: Dependabot PRs have restricted permissions for testing
Per [this link](https://docs.github.com/en/code-security/dependabot/working-with-dependabot/automating-dependabot-with-github-actions#changing-github_token-permissions):

> By default, GitHub Actions workflows triggered by Dependabot get a GITHUB_TOKEN with read-only permissions. You can use the permissions key in your workflow to increase the access for the token

Additionally, dependabot has a [separate set of available secrets](https://docs.github.com/en/code-security/dependabot/working-with-dependabot/configuring-access-to-private-registries-for-dependabot#about-encrypted-secrets-for-dependabot):

> Dependabot secrets also include secrets that are used by GitHub Actions workflows triggered by Dependabot pull requests. Dependabot itself may not use these secrets, but the workflows require them.

That means that this construct to use `secrets.UCLAHS_CDS_REPO_READ_TOKEN`...

https://github.com/uclahs-cds/tool-Nextflow-action/blob/da23bc9d87fda1d5b6751d3b671057d0b4d7e297/.github/workflows/nextflow-tests.yml#L245-L255

... _does not_ work on dependabot PRs.

### LINKED ISSUE #32: Handle review dismissal protections
Using https://github.com/uclahs-cds/pipeline-align-DNA/actions/runs/8542074596/attempts/1 as an example, the tests passed but the bot's own review could not be dismissed:

>  RequestError [HttpError]: Branch protections do not permit dismissing this review

Per https://github.blog/2017-03-06-restrict-review-dismissals-with-protected-branches/, that's because we have the branch protection rule set to restrict who can dismiss pull request reviews:
Screenshot 2024-04-03 at 9 40 15 AM

So the bot can _create_ reviews, but it cannot _dismiss_ them.

I specifically didn't want the bot to approve PRs, as I thought that would mess with the required number of approvals before merging. As I see it there are a few paths forward:

1. Continue to request changes, but pass an administrator PAT via `secrets` and use that to dismiss the review. That runs into similar issues as #31.
2. Continue to request changes, but approve after tests pass. Maybe GitHub has some special logic that workflow approvals don't count like human approvals for purposes of branch protections? Maybe?
3. Abandon reviews and use comments to communicate with the user. I was using non-dismissed reviews as a state machine to avoid re-posting the same comment or attempting to fix non-broken tests, but I could theoretically do the same thing with comments.

### PR DESCRIPTION
**Pull request by nwiltsie**
# Description

This is a major reworking of the Nextflow testing workflow that fixes #30, fixes #31, and fixes #32.

You can see a full demo of the updated workflow in https://github.com/uclahs-cds/user-nwiltsie-pipeline/pull/18 (_and_ the success after merging, [here](https://github.com/uclahs-cds/user-nwiltsie-pipeline/actions/runs/8574441766/job/23501273394) and [here](https://github.com/uclahs-cds/user-nwiltsie-pipeline/actions/runs/8574444159)).

The current workflow, as of #29, has a few problems:

* PRs opened by dependabot have restricted permissions (#31)
* Reviews on PRs to protected branches can only be dismissed by maintainers (#32)
* As running the tests inherently involves executing pipeline code, we are theoretically vulnerable to maliciously-crafted pipelines that could access organizational secrets

Following advice from [this link](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/), I've now split the test execution into two workflow: a low-privilege workflow that performs a bunch of safety checks and establishes a common baseline, and a high-privilege workflow that triggers on the successful conclusion of the first.

That means that each pipeline will need two workflows like this:

```yaml
---
name: Trigger tests
on:
  issue_comment:
    types: [created]
  pull_request_target:
    branches:
      - main
  push:
    branches:
      - main

permissions:
  pull-requests: write
  statuses: write

jobs:
  check-user:
    uses: uclahs-cds/tool-Nextflow-action/.github/workflows/test-setup.yml@main
```

```yaml
---
name: Nextflow

on:
  workflow_run:
    workflows: [Trigger Tests]
    types:
      - completed

permissions:
  pull-requests: write
  contents: write
  packages: read
  actions: read
  statuses: write

jobs:
  tests:
    if: ${{ github.event.workflow_run.conclusion == 'success' }}
    uses: uclahs-cds/tool-Nextflow-action/.github/workflows/nextflow-tests.yml@main
    secrets: inherit
 ```

The first job protects against the following things:
* Users without write permissions trying to run tests
* Race conditions (privileged users triggering tests and malicious users _immediately_ pushing new changes)
* Fixes not being possible

out

The second job has more elevated permissions, but will only run after the success of the first.

---

Unfortunately, this is a breaking change with the current implementation of the tests. Here are the next steps:

1. Get this PR approved
2. Disable testing workflow in each existing pipeline
3. Merge these changes
4. Open PRs to update workflows in existing pipelines

# Checklist

- [x] This PR **does *NOT* contain** Protected Health Information [(PHI)](https://ohrpp.research.ucla.edu/hipaa/). A repo may ***need to be deleted*** if such data is uploaded. <br> Disclosing PHI is a ***major problem***[^1] - Even ***a small leak can be costly***[^2].
  
- [x] This PR **does *NOT* contain** germline genetic data[^3], RNA-Seq, DNA methylation, microbiome or other molecular data[^4].

[^1]: [UCLA Health reaches $7.5m settlement over 2015 breach of 4.5m patient records](https://healthitsecurity.com/news/ucla-health-reaches-7.5m-settlement-over-2015-breach-of-4.5m)
[^2]: [The average healthcare data breach costs $2.2 million, despite the majority of breaches releasing fewer than 500 records.](https://www.ponemon.org/local/upload/file/Sixth%20Annual%20Patient%20Privacy%20%26%20Data%20Security%20Report%20FINAL%206.pdf)
[^3]: [Genetic information is considered PHI.](https://www.genome.gov/about-genomics/policy-issues/Privacy#:~:text=In%202013%2C%20as%20required%20by,genetic%20information%20for%20underwriting%20purposes.)
  [Forensic assays can identify patients with as few as 21 SNPs](https://www.sciencedirect.com/science/article/pii/S1525157817305962)
[^4]: [RNA-Seq](https://www.nature.com/articles/ng.2248), [DNA methylation](https://ieeexplore.ieee.org/document/7958619), [microbiome](https://www.pnas.org/doi/pdf/10.1073/pnas.1423854112), or other molecular data can be used to predict genotypes (PHI) and reveal a patient's identity.

- [x] This PR **does *NOT* contain** other non-plain text files, such as: compressed files, images (*e.g.* `.png`, .`jpeg`), `.pdf`, `.RData`, `.xlsx`, `.doc`, `.ppt`, or other output files.

_&emsp; To automatically exclude such files using a [.gitignore](https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files) file, see [here](https://github.com/uclahs-cds/template-base/blob/main/.gitignore) for example._

- [x] I have read the [code review guidelines](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3187646/Code+Review+Guidelines) and the [code review best practice on GitHub check-list](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3189956/Code+Review+Best+Practice+on+GitHub+-+Check+List).

- [x] I have set up or verified the `main` branch protection rule following the [github standards](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3190380/GitHub+Standards#GitHubStandards-Branchprotectionrule) before opening this pull request.

- [x] The name of the branch is meaningful and well formatted following the [standards](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3189956/Code+Review+Best+Practice+on+GitHub+-+Check+List), using [AD_username (or 5 letters of AD if AD is too long)]-[brief_description_of_branch].
  
- [x] I have added the major changes included in this pull request to the `CHANGELOG.md` under the next release version or unreleased, and updated the date.

### CODE REVIEW

**Comment by nwiltsie** [comment]
A few other details:

Reviews on protected branches can't be dismissed: _fine_, the bot just uses comments now.

In addition to `/fix-tests`, commenting `/run-tests` will (re-)run the tests, as if new changes were pushed. That command is also protected against users without write access to the repo.

Due to using `workflow_run`, the test jobs don't get automatic status checks on the PR anymore. The low-permission/filtering job _does_ show up. I'm adding an explicit `Nextflow tests` status, updated three times:

1. End of the low-permission job: set to `pending` with no link
2. Start of the high-permission job: set to `pending` with the link to the job
3. End of the high-permission job: set to `success` or `failure`

Screenshot 2024-04-05 at 12 44 01 PM

File: .github/workflows/test-setup.yml [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+                });
+
+                core.setFailed(`${triggering_actor} does not have write permissions`);
+            }
+
+      - uses: actions/github-script@v7
+        name: Prep for test
+        with:
+          script: |
+            // Create an archive with all of the relevant control information
```
Comment (aholmes) [inline .github/workflows/test-setup.yml:96]: Can github-script use external files for the scripts? We may want to consider moving them out of the YAML file for ease of maintenance and testability.

Comment (nwiltsie) [reply .github/workflows/test-setup.yml:96]: I'm 100% with you, and in general I think _no_ - I'll look for more details, but I don't think the reusable workflow repo gets checked out.

Comment (nwiltsie) [reply .github/workflows/test-setup.yml:96]: It seems like we could - https://github.com/orgs/community/discussions/25294 - that might be worth it.

Comment (nwiltsie) [reply .github/workflows/test-setup.yml:96]: Also https://github.com/orgs/community/discussions/25294

**Review by aholmes — APPROVED** [review]

**Pull request merged**